File: frontend/src/mocks/factories/model-monitoring.ts
/**
 * Factory for generating mock model monitoring data.
 *
 * Provides realistic ML model monitoring data including:
 * - Registered models with status and metrics
 * - Alert rules and handlers
 * - Alert instances
 * - Performance metrics time series
 */

import { faker } from '@faker-js/faker'
import { createId, createTimestamp, randomChoice } from './base'

// ============================================================================
// Types
// ============================================================================

export interface RegisteredModel {
  id: string
  name: string
  version: string
  description: string
  status: 'active' | 'paused' | 'degraded' | 'error'
  config: {
    enable_drift_detection: boolean
    enable_quality_metrics: boolean
    enable_performance_metrics: boolean
    sample_rate: number
    drift_threshold: number
    drift_window_size: number
  }
  metadata: Record<string, string>
  prediction_count: number
  last_prediction_at: string | null
  current_drift_score: number | null
  health_score: number
  created_at: string
  updated_at: string
}

export interface AlertRule {
  id: string
  name: string
  model_id: string
  rule_type: 'threshold' | 'statistical' | 'trend'
  severity: 'critical' | 'warning' | 'info'
  config: {
    metric_name?: string
    threshold?: number
    comparison?: 'gt' | 'lt' | 'gte' | 'lte' | 'eq'
    duration_seconds?: number
    std_deviations?: number
    trend_direction?: 'up' | 'down'
    trend_period_hours?: number
  }
  is_active: boolean
  last_triggered_at: string | null
  trigger_count: number
  created_at: string
  updated_at: string
}

export interface AlertHandler {
  id: string
  name: string
  handler_type: 'slack' | 'webhook' | 'email'
  config: {
    webhook_url?: string
    channel?: string
    username?: string
    url?: string
    method?: string
    headers?: Record<string, string>
    recipients?: string[]
    from_address?: string
    subject_template?: string
  }
  is_active: boolean
  send_count: number
  failure_count: number
  last_sent_at: string | null
  last_error: string | null
  created_at: string
  updated_at: string
}

export interface AlertInstance {
  id: string
  rule_id: string
  model_id: string
  model_name: string
  severity: 'critical' | 'warning' | 'info'
  message: string
  metric_value: number | null
  threshold_value: number | null
  acknowledged: boolean
  acknowledged_by: string | null
  acknowledged_at: string | null
  resolved: boolean
  resolved_at: string | null
  created_at: string
  updated_at: string
}

export interface MetricDataPoint {
  timestamp: string
  value: number
}

export interface MetricSummary {
  name: string
  type: string
  count: number
  min_value: number | null
  max_value: number | null
  avg_value: number | null
  p50_value: number | null
  p95_value: number | null
  p99_value: number | null
  last_value: number | null
}

// ============================================================================
// Model Names
// ============================================================================

const MODEL_NAMES = [
  'fraud-detection',
  'churn-prediction',
  'recommendation-engine',
  'sentiment-analyzer',
  'price-predictor',
  'image-classifier',
  'text-generator',
  'anomaly-detector',
  'risk-scorer',
  'demand-forecaster',
]

// ============================================================================
// Factory Functions
// ============================================================================

export function createTimestamps() {
  const created = createTimestamp(faker.number.int({ min: 1, max: 90 }))
  return {
    created_at: created,
    updated_at: faker.date.between({ from: created, to: new Date() }).toISOString(),
  }
}

export function createRegisteredModel(overrides: Partial<RegisteredModel> = {}): RegisteredModel {
  const timestamps = createTimestamps()
  const status = randomChoice(['active', 'active', 'active', 'paused', 'degraded'] as const)
  const healthScore =
    status === 'active'
      ? faker.number.float({ min: 70, max: 100, fractionDigits: 1 })
      : status === 'degraded'
        ? faker.number.float({ min: 40, max: 69, fractionDigits: 1 })
        : faker.number.float({ min: 50, max: 90, fractionDigits: 1 })

  return {
    id: createId(),
    name: randomChoice(MODEL_NAMES),
    version: `${faker.number.int({ min: 1, max: 3 })}.${faker.number.int({ min: 0, max: 9 })}.${faker.number.int({ min: 0, max: 9 })}`,
    description: faker.lorem.sentence(),
    status,
    config: {
      enable_drift_detection: true,
      enable_quality_metrics: true,
      enable_performance_metrics: true,
      sample_rate: faker.helpers.arrayElement([0.1, 0.25, 0.5, 1.0]),
      drift_threshold: faker.helpers.arrayElement([0.05, 0.1, 0.15]),
      drift_window_size: faker.helpers.arrayElement([500, 1000, 2000, 5000]),
    },
    metadata: {
      team: faker.helpers.arrayElement(['ml-platform', 'data-science', 'analytics']),
      owner: faker.person.fullName(),
    },
    prediction_count: faker.number.int({ min: 1000, max: 5000000 }),
    last_prediction_at: faker.date.recent({ days: 1 }).toISOString(),
    current_drift_score: faker.helpers.maybe(
      () => faker.number.float({ min: 0, max: 0.3, fractionDigits: 3 }),
      { probability: 0.8 }
    ) ?? null,
    health_score: healthScore,
    ...timestamps,
    ...overrides,
  }
}

export function createAlertRule(
  modelId: string,
  overrides: Partial<AlertRule> = {}
): AlertRule {
  const timestamps = createTimestamps()
  const ruleType = randomChoice(['threshold', 'statistical', 'trend'] as const)
  const severity = randomChoice(['critical', 'warning', 'info'] as const)

  let config: AlertRule['config'] = {}
  if (ruleType === 'threshold') {
    config = {
      metric_name: randomChoice(['latency_ms', 'error_rate', 'null_rate', 'drift_score']),
      threshold: faker.number.float({ min: 50, max: 500, fractionDigits: 1 }),
      comparison: randomChoice(['gt', 'lt', 'gte', 'lte'] as const),
      duration_seconds: faker.helpers.arrayElement([60, 300, 600]),
    }
  } else if (ruleType === 'statistical') {
    config = {
      metric_name: randomChoice(['latency_ms', 'throughput']),
      std_deviations: faker.helpers.arrayElement([2, 3, 4]),
    }
  } else {
    config = {
      metric_name: randomChoice(['latency_ms', 'error_rate']),
      trend_direction: randomChoice(['up', 'down'] as const),
      trend_period_hours: faker.helpers.arrayElement([1, 6, 24]),
    }
  }

  const triggerCount = faker.number.int({ min: 0, max: 50 })

  return {
    id: createId(),
    name: `${severity === 'critical' ? 'Critical' : severity === 'warning' ? 'Warning' : 'Info'} ${config.metric_name} Alert`,
    model_id: modelId,
    rule_type: ruleType,
    severity,
    config,
    is_active: faker.datatype.boolean({ probability: 0.8 }),
    last_triggered_at:
      triggerCount > 0 ? faker.date.recent({ days: 7 }).toISOString() : null,
    trigger_count: triggerCount,
    ...timestamps,
    ...overrides,
  }
}

export function createAlertHandler(overrides: Partial<AlertHandler> = {}): AlertHandler {
  const timestamps = createTimestamps()
  const handlerType = randomChoice(['slack', 'webhook', 'email'] as const)

  let config: AlertHandler['config'] = {}
  if (handlerType === 'slack') {
    config = {
      webhook_url: `https://hooks.slack.com/services/${faker.string.alphanumeric(9)}/${faker.string.alphanumeric(11)}/${faker.string.alphanumeric(24)}`,
      channel: `#${randomChoice(['alerts', 'ml-monitoring', 'ops', 'data-quality'])}`,
      username: 'Truthound Bot',
    }
  } else if (handlerType === 'webhook') {
    config = {
      url: `https://${faker.internet.domainName()}/webhook/alerts`,
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${faker.string.alphanumeric(32)}`,
      },
    }
  } else {
    config = {
      recipients: [faker.internet.email(), faker.internet.email()],
      from_address: `alerts@${faker.internet.domainName()}`,
      subject_template: '[{{severity}}] Model {{model_name}}: {{alert_name}}',
    }
  }

  const sendCount = faker.number.int({ min: 0, max: 200 })
  const failureCount = faker.number.int({ min: 0, max: Math.floor(sendCount * 0.1) })

  return {
    id: createId(),
    name: `${handlerType.charAt(0).toUpperCase() + handlerType.slice(1)} Handler - ${faker.company.buzzNoun()}`,
    handler_type: handlerType,
    config,
    is_active: faker.datatype.boolean({ probability: 0.85 }),
    send_count: sendCount,
    failure_count: failureCount,
    last_sent_at: sendCount > 0 ? faker.date.recent({ days: 3 }).toISOString() : null,
    last_error: failureCount > 0 && faker.datatype.boolean({ probability: 0.3 })
      ? 'Connection timeout'
      : null,
    ...timestamps,
    ...overrides,
  }
}

export function createAlertInstance(
  ruleId: string,
  modelId: string,
  modelName: string,
  overrides: Partial<AlertInstance> = {}
): AlertInstance {
  const timestamps = createTimestamps()
  const severity = randomChoice(['critical', 'warning', 'info'] as const)
  const acknowledged = faker.datatype.boolean({ probability: 0.6 })
  const resolved = acknowledged && faker.datatype.boolean({ probability: 0.5 })

  return {
    id: createId(),
    rule_id: ruleId,
    model_id: modelId,
    model_name: modelName,
    severity,
    message: faker.helpers.arrayElement([
      `Latency exceeded threshold: ${faker.number.float({ min: 100, max: 500, fractionDigits: 1 })}ms > 100ms`,
      `Drift detected: score ${faker.number.float({ min: 0.1, max: 0.4, fractionDigits: 2 })} > 0.1`,
      `Error rate spike: ${faker.number.float({ min: 5, max: 20, fractionDigits: 1 })}%`,
      `Null rate increased: ${faker.number.float({ min: 2, max: 10, fractionDigits: 1 })}%`,
      `Throughput dropped by ${faker.number.int({ min: 20, max: 50 })}%`,
    ]),
    metric_value: faker.number.float({ min: 50, max: 500, fractionDigits: 2 }),
    threshold_value: faker.number.float({ min: 50, max: 100, fractionDigits: 2 }),
    acknowledged,
    acknowledged_by: acknowledged ? faker.person.firstName() : null,
    acknowledged_at: acknowledged
      ? faker.date.between({ from: timestamps.created_at, to: new Date() }).toISOString()
      : null,
    resolved,
    resolved_at: resolved ? faker.date.recent({ days: 1 }).toISOString() : null,
    ...timestamps,
    ...overrides,
  }
}

export function createMetricDataPoints(
  metricName: string,
  hours: number = 24
): MetricDataPoint[] {
  const points: MetricDataPoint[] = []
  const now = new Date()
  const interval = 30 // 30 minutes
  const numPoints = Math.min((hours * 60) / interval, 100)

  let baseValue: number
  let variance: number

  switch (metricName) {
    case 'latency_ms':
      baseValue = faker.number.float({ min: 30, max: 60 })
      variance = 20
      break
    case 'throughput':
      baseValue = faker.number.float({ min: 100, max: 300 })
      variance = 50
      break
    case 'error_rate':
      baseValue = faker.number.float({ min: 0.5, max: 2 })
      variance = 1
      break
    default:
      baseValue = 50
      variance = 10
  }

  for (let i = numPoints - 1; i >= 0; i--) {
    const timestamp = new Date(now.getTime() - i * interval * 60 * 1000)
    // Add some sinusoidal pattern for realistic data
    const sinComponent = Math.sin((i / numPoints) * Math.PI * 4) * variance * 0.3
    const randomComponent = (Math.random() - 0.5) * variance
    const value = Math.max(0, baseValue + sinComponent + randomComponent)

    points.push({
      timestamp: timestamp.toISOString(),
      value: parseFloat(value.toFixed(2)),
    })
  }

  return points
}

export function createMetricSummary(dataPoints: MetricDataPoint[]): MetricSummary {
  if (dataPoints.length === 0) {
    return {
      name: '',
      type: '',
      count: 0,
      min_value: null,
      max_value: null,
      avg_value: null,
      p50_value: null,
      p95_value: null,
      p99_value: null,
      last_value: null,
    }
  }

  const values = dataPoints.map((p) => p.value)
  const sorted = [...values].sort((a, b) => a - b)
  const n = sorted.length

  return {
    name: '',
    type: '',
    count: n,
    min_value: sorted[0],
    max_value: sorted[n - 1],
    avg_value: values.reduce((a, b) => a + b, 0) / n,
    p50_value: sorted[Math.floor(n * 0.5)],
    p95_value: sorted[Math.floor(n * 0.95)],
    p99_value: sorted[Math.floor(n * 0.99)],
    last_value: values[n - 1],
  }
}

// ============================================================================
// Batch Creation
// ============================================================================

export function createDiverseModels(count: number = 5): RegisteredModel[] {
  const usedNames = new Set<string>()
  const models: RegisteredModel[] = []

  for (let i = 0; i < count; i++) {
    let name: string
    do {
      name = randomChoice(MODEL_NAMES)
    } while (usedNames.has(name) && usedNames.size < MODEL_NAMES.length)
    usedNames.add(name)

    models.push(createRegisteredModel({ name }))
  }

  return models
}

export function createDiverseAlertRules(modelIds: string[]): AlertRule[] {
  const rules: AlertRule[] = []

  modelIds.forEach((modelId) => {
    // Create 2-4 rules per model
    const ruleCount = faker.number.int({ min: 2, max: 4 })
    for (let i = 0; i < ruleCount; i++) {
      rules.push(createAlertRule(modelId))
    }
  })

  return rules
}

export function createDiverseAlertHandlers(count: number = 4): AlertHandler[] {
  const handlers: AlertHandler[] = []
  const types: Array<'slack' | 'webhook' | 'email'> = ['slack', 'webhook', 'email']

  // Ensure at least one of each type
  types.forEach((type) => {
    handlers.push(
      createAlertHandler({
        handler_type: type,
        name: `Primary ${type.charAt(0).toUpperCase() + type.slice(1)} Handler`,
      })
    )
  })

  // Add extra handlers
  for (let i = handlers.length; i < count; i++) {
    handlers.push(createAlertHandler())
  }

  return handlers
}

export function createDiverseAlertInstances(
  rules: AlertRule[],
  models: RegisteredModel[],
  count: number = 10
): AlertInstance[] {
  const alerts: AlertInstance[] = []
  const modelMap = new Map(models.map((m) => [m.id, m]))

  for (let i = 0; i < count; i++) {
    const rule = randomChoice(rules)
    const model = modelMap.get(rule.model_id)
    if (model) {
      alerts.push(createAlertInstance(rule.id, model.id, model.name))
    }
  }

  return alerts
}

File: src/truthound_dashboard/schemas/model_monitoring.py
"""Pydantic schemas for ML Model Monitoring.

Provides schemas for:
- Model registration and management
- Performance metrics (latency, throughput, error rates)
- Data quality metrics (null rates, type violations)
- Drift detection integration
- Alert rules and handlers
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any, Literal

from pydantic import BaseModel, Field

from .base import BaseSchema, IDMixin, ListResponseWrapper, TimestampMixin


# =============================================================================
# Enums
# =============================================================================


class ModelStatus(str, Enum):
    """Model monitoring status."""

    ACTIVE = "active"
    PAUSED = "paused"
    DEGRADED = "degraded"
    ERROR = "error"


class AlertSeverity(str, Enum):
    """Alert severity levels."""

    CRITICAL = "critical"
    WARNING = "warning"
    INFO = "info"


class MetricType(str, Enum):
    """Types of metrics collected."""

    LATENCY = "latency"
    THROUGHPUT = "throughput"
    ERROR_RATE = "error_rate"
    NULL_RATE = "null_rate"
    TYPE_VIOLATION = "type_violation"
    DRIFT_SCORE = "drift_score"
    CUSTOM = "custom"


class AlertRuleType(str, Enum):
    """Types of alert rules.

    Maps to truthound.ml.monitoring.alerting rules:
    - ThresholdRule: Threshold-based alerting
    - AnomalyRule: Anomaly-based alerting (statistical)
    - TrendRule: Trend-based alerting
    """

    THRESHOLD = "threshold"
    STATISTICAL = "statistical"  # Maps to AnomalyRule
    TREND = "trend"


class AlertHandlerType(str, Enum):
    """Types of alert handlers.

    Maps to truthound.ml.monitoring.alerting handlers:
    - SlackAlertHandler
    - PagerDutyAlertHandler
    - WebhookAlertHandler
    """

    SLACK = "slack"
    WEBHOOK = "webhook"
    EMAIL = "email"
    PAGERDUTY = "pagerduty"


# =============================================================================
# Model Registration Schemas
# =============================================================================


class ModelConfigBase(BaseSchema):
    """Configuration for model monitoring.

    Maps to truthound.ml.monitoring.MonitorConfig parameters.
    See: .truthound_docs/advanced/ml-anomaly.md#model-monitoring
    """

    # Core monitoring settings (from MonitorConfig)
    batch_size: int = Field(
        default=100,
        description="Batch size for metric collection",
        ge=1,
        le=10000,
    )
    collect_interval_seconds: int = Field(
        default=60,
        description="Interval for collecting metrics (seconds)",
        ge=1,
        le=3600,
    )
    alert_evaluation_interval_seconds: int = Field(
        default=30,
        description="Interval for evaluating alert rules (seconds)",
        ge=1,
        le=3600,
    )
    retention_hours: int = Field(
        default=24,
        description="Hours to retain metrics data",
        ge=1,
        le=720,  # 30 days max
    )

    # Feature toggles
    enable_drift_detection: bool = Field(
        default=True,
        description="Enable drift detection using th.compare()",
    )
    enable_quality_metrics: bool = Field(
        default=True,
        description="Enable quality metrics (accuracy, precision, recall, F1)",
    )
    enable_performance_metrics: bool = Field(
        default=True,
        description="Enable performance metrics (latency, throughput, error rate)",
    )

    # Drift detection settings
    drift_threshold: float = Field(
        default=0.1,
        description="Threshold for drift detection alerts",
        ge=0.0,
        le=1.0,
    )
    drift_method: str = Field(
        default="auto",
        description="Drift detection method (auto, psi, ks, js, wasserstein, chi2, etc.)",
    )


class RegisteredModelBase(BaseSchema):
    """Base schema for registered model."""

    name: str = Field(..., description="Model name/identifier", min_length=1, max_length=255)
    version: str = Field(default="1.0.0", description="Model version")
    description: str = Field(default="", description="Model description")
    status: ModelStatus = Field(default=ModelStatus.ACTIVE, description="Monitoring status")
    config: ModelConfigBase = Field(default_factory=ModelConfigBase, description="Monitoring config")
    metadata: dict[str, Any] = Field(default_factory=dict, description="Additional metadata")


class RegisterModelRequest(RegisteredModelBase):
    """Request to register a model for monitoring."""

    pass


class UpdateModelRequest(BaseSchema):
    """Request to update model registration."""

    name: str | None = Field(None, description="Model name")
    version: str | None = Field(None, description="Model version")
    description: str | None = Field(None, description="Model description")
    status: ModelStatus | None = Field(None, description="Monitoring status")
    config: ModelConfigBase | None = Field(None, description="Monitoring config")
    metadata: dict[str, Any] | None = Field(None, description="Additional metadata")


class RegisteredModelResponse(RegisteredModelBase, IDMixin, TimestampMixin):
    """Response for registered model."""

    prediction_count: int = Field(default=0, description="Total predictions")
    last_prediction_at: datetime | None = Field(None, description="Last prediction timestamp")
    current_drift_score: float | None = Field(None, description="Current drift score")
    health_score: float = Field(default=100.0, description="Model health score (0-100)")


class RegisteredModelListResponse(ListResponseWrapper):
    """List response for registered models."""

    items: list[RegisteredModelResponse]


# =============================================================================
# Metrics Schemas
# =============================================================================


class MetricDataPoint(BaseModel):
    """Single metric data point."""

    timestamp: datetime = Field(..., description="Measurement timestamp")
    value: float = Field(..., description="Metric value")
    labels: dict[str, str] = Field(default_factory=dict, description="Metric labels")


class MetricSummary(BaseModel):
    """Summary statistics for a metric."""

    name: str = Field(..., description="Metric name")
    type: MetricType = Field(..., description="Metric type")
    count: int = Field(default=0, description="Number of observations")
    min_value: float | None = Field(None, description="Minimum value")
    max_value: float | None = Field(None, description="Maximum value")
    avg_value: float | None = Field(None, description="Average value")
    p50_value: float | None = Field(None, description="50th percentile")
    p95_value: float | None = Field(None, description="95th percentile")
    p99_value: float | None = Field(None, description="99th percentile")
    last_value: float | None = Field(None, description="Most recent value")


class RecordPredictionRequest(BaseModel):
    """Request to record a model prediction."""

    features: dict[str, Any] = Field(..., description="Input features")
    prediction: Any = Field(..., description="Model prediction output")
    actual: Any | None = Field(None, description="Actual value (for accuracy tracking)")
    latency_ms: float | None = Field(None, description="Prediction latency in ms", ge=0)
    metadata: dict[str, Any] = Field(default_factory=dict, description="Additional metadata")


class RecordPredictionResponse(BaseModel):
    """Response for recorded prediction."""

    id: str = Field(..., description="Prediction ID")
    model_id: str = Field(..., description="Model ID")
    recorded_at: datetime = Field(..., description="Recording timestamp")


class MetricsResponse(BaseModel):
    """Response containing model metrics."""

    model_id: str = Field(..., description="Model ID")
    model_name: str = Field(..., description="Model name")
    time_range_hours: int = Field(default=24, description="Time range for metrics")
    metrics: list[MetricSummary] = Field(default_factory=list, description="Metric summaries")
    data_points: dict[str, list[MetricDataPoint]] = Field(
        default_factory=dict,
        description="Time series data points by metric name",
    )


# =============================================================================
# Alert Rule Schemas
# =============================================================================


class ThresholdRuleConfig(BaseModel):
    """Configuration for threshold-based alert rule."""

    metric_name: str = Field(..., description="Metric to monitor")
    threshold: float = Field(..., description="Threshold value")
    comparison: Literal["gt", "lt", "gte", "lte", "eq"] = Field(
        default="gt",
        description="Comparison operator",
    )
    duration_seconds: int = Field(
        default=0,
        description="Condition must persist for this duration",
        ge=0,
    )


class StatisticalRuleConfig(BaseModel):
    """Configuration for statistical alert rule."""

    metric_name: str = Field(..., description="Metric to monitor")
    std_devs: float = Field(default=3.0, description="Standard deviations for anomaly", ge=1.0)
    window_size: int = Field(default=100, description="Window size for baseline", ge=10)


class AlertRuleBase(BaseSchema):
    """Base schema for alert rules."""

    name: str = Field(..., description="Rule name", min_length=1, max_length=255)
    model_id: str = Field(..., description="Model ID this rule applies to")
    rule_type: AlertRuleType = Field(..., description="Type of alert rule")
    severity: AlertSeverity = Field(default=AlertSeverity.WARNING, description="Alert severity")
    config: dict[str, Any] = Field(..., description="Rule-specific configuration")
    is_active: bool = Field(default=True, description="Whether rule is active")


class CreateAlertRuleRequest(AlertRuleBase):
    """Request to create an alert rule."""

    pass


class UpdateAlertRuleRequest(BaseSchema):
    """Request to update an alert rule."""

    name: str | None = Field(None, description="Rule name")
    severity: AlertSeverity | None = Field(None, description="Alert severity")
    config: dict[str, Any] | None = Field(None, description="Rule configuration")
    is_active: bool | None = Field(None, description="Whether rule is active")


class AlertRuleResponse(AlertRuleBase, IDMixin, TimestampMixin):
    """Response for alert rule."""

    last_triggered_at: datetime | None = Field(None, description="Last trigger time")
    trigger_count: int = Field(default=0, description="Total triggers")


class AlertRuleListResponse(ListResponseWrapper):
    """List response for alert rules."""

    items: list[AlertRuleResponse]


# =============================================================================
# Alert Handler Schemas
# =============================================================================


class SlackHandlerConfig(BaseModel):
    """Configuration for Slack alert handler."""

    webhook_url: str = Field(..., description="Slack webhook URL")
    channel: str | None = Field(None, description="Override channel")
    username: str = Field(default="Truthound", description="Bot username")


class WebhookHandlerConfig(BaseModel):
    """Configuration for webhook alert handler."""

    url: str = Field(..., description="Webhook URL")
    method: Literal["POST", "PUT"] = Field(default="POST", description="HTTP method")
    headers: dict[str, str] = Field(default_factory=dict, description="Custom headers")


class AlertHandlerBase(BaseSchema):
    """Base schema for alert handlers."""

    name: str = Field(..., description="Handler name", min_length=1, max_length=255)
    handler_type: AlertHandlerType = Field(..., description="Type of handler")
    config: dict[str, Any] = Field(..., description="Handler-specific configuration")
    is_active: bool = Field(default=True, description="Whether handler is active")


class CreateAlertHandlerRequest(AlertHandlerBase):
    """Request to create an alert handler."""

    pass


class UpdateAlertHandlerRequest(BaseSchema):
    """Request to update an alert handler."""

    name: str | None = Field(None, description="Handler name")
    config: dict[str, Any] | None = Field(None, description="Handler configuration")
    is_active: bool | None = Field(None, description="Whether handler is active")


class AlertHandlerResponse(AlertHandlerBase, IDMixin, TimestampMixin):
    """Response for alert handler."""

    last_sent_at: datetime | None = Field(None, description="Last alert sent time")
    send_count: int = Field(default=0, description="Total alerts sent")
    failure_count: int = Field(default=0, description="Total failures")


class AlertHandlerListResponse(ListResponseWrapper):
    """List response for alert handlers."""

    items: list[AlertHandlerResponse]


# =============================================================================
# Alert Instance Schemas
# =============================================================================


class AlertInstance(BaseModel, TimestampMixin):
    """An instance of a triggered alert."""

    id: str = Field(..., description="Alert instance ID")
    rule_id: str = Field(..., description="Rule that triggered")
    model_id: str = Field(..., description="Model ID")
    severity: AlertSeverity = Field(..., description="Alert severity")
    message: str = Field(..., description="Alert message")
    metric_value: float | None = Field(None, description="Metric value that triggered")
    threshold_value: float | None = Field(None, description="Threshold value")
    acknowledged: bool = Field(default=False, description="Whether acknowledged")
    acknowledged_by: str | None = Field(None, description="Who acknowledged")
    acknowledged_at: datetime | None = Field(None, description="When acknowledged")
    resolved: bool = Field(default=False, description="Whether resolved")
    resolved_at: datetime | None = Field(None, description="When resolved")


class AlertListResponse(ListResponseWrapper):
    """List response for alerts."""

    items: list[AlertInstance]


class AcknowledgeAlertRequest(BaseModel):
    """Request to acknowledge an alert."""

    actor: str = Field(..., description="Who is acknowledging")


# =============================================================================
# Dashboard Schemas
# =============================================================================


class ModelDashboardData(BaseModel):
    """Dashboard data for a single model."""

    model: RegisteredModelResponse = Field(..., description="Model information")
    metrics: MetricsResponse = Field(..., description="Current metrics")
    active_alerts: list[AlertInstance] = Field(default_factory=list, description="Active alerts")
    recent_predictions: int = Field(default=0, description="Predictions in last hour")
    health_status: str = Field(default="healthy", description="Overall health status")


class MonitoringOverview(BaseModel):
    """Overview of all monitored models."""

    total_models: int = Field(default=0, description="Total registered models")
    active_models: int = Field(default=0, description="Models actively receiving predictions")
    degraded_models: int = Field(default=0, description="Models in degraded state")
    total_predictions_24h: int = Field(default=0, description="Total predictions in 24h")
    active_alerts: int = Field(default=0, description="Currently active alerts")
    models_with_drift: int = Field(default=0, description="Models with detected drift")
    avg_latency_ms: float | None = Field(None, description="Average latency across all models")


class AlertHandlerTestResult(BaseModel):
    """Result of testing an alert handler."""

    success: bool = Field(..., description="Whether test was successful")
    message: str = Field(..., description="Test result message")
    handler_id: str = Field(..., description="Handler ID that was tested")
    handler_type: str = Field(..., description="Type of the handler")


class RuleEvaluationResult(BaseModel):
    """Result of rule evaluation for a model."""

    model_id: str = Field(..., description="Model ID that was evaluated")
    alerts_created: int = Field(..., description="Number of alerts created")
    alert_ids: list[str] = Field(default_factory=list, description="IDs of created alerts")
